compiler/src/dotty/tools/dotc/transform/init/ParamOverridingCheck.scala

```diff
@@ -0,0 +1,343 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import core.*
+import Contexts.*
+import Symbols.*
+import Types.*
+import StdNames.*
+import NameKinds.OuterSelectName
+import util.Property
+
+import ast.tpd.*
+import config.Printers.init as printer
+import reporting.trace as log
+
+import Semantic.Arg
+import Semantic.NewExpr
+import Semantic.Call
+import Semantic.hasSource
+
+import scala.collection.mutable
+import scala.annotation.tailrec
+
+/**
+ * Check overriding of class parameters
+ *
+ * This check issues a warning if the use of a class pamameter in the primary
+ * constructor potentially has different semantics from its use in methods.
+ * The subtle semantic difference is demonstrated by the following exmpale:
+ *
+ *     class B(val y: Int):
+ *       println(y)                   // 10
+ *       foo()
+ *       def foo() = println(y)       // 20
+ *
+ *     class C(override val y: Int) extends B(10)
+ *
+ *     new C(20)
+ *
+ * A well-formed program should not depend on such subtle semantic differences.
+ * Therefore, we detect and warn such subtle semantic differences in code.
+ *
+ * This check depends on loading TASTY from libraries. It can be enabled with
+ * the compiler option `-Ysafe-init`.
+ */
+object ParamOverridingCheck:
+  type Contextual[T] = Context ?=> T
+
+  private val IsInSuperCall = new Property.Key[Boolean]
+
+  // ----------------------- Domain definitions --------------------------------
+
+  sealed abstract class Value:
+    val source: Tree
+
+  /**
+   * A symbolic value
+   *
+   * Equality is defined as referential equality.
+   */
+  class Skolem(val source: Tree) extends Value
+
```

Comment:
Or do you care about two different instances of the same `Skolem`? Can you guarantee that both will evaluate to the same concrete value?

Reply:
I'm getting confused. If you pass `Random.nextInt()` for two different parameters, then pass these parameters respectively to the superclass parameter and the overriding parameter, will we avoid a warning because `Random.nextInt()` and `Random.nextInt()` are the same trees, even though they evaluate to different values?

Reply:
Different instances of `Random.nextInt()` will be different `Skolem`s. The same Skolem matters when they are passed as arguments so that we can reason about parameter equality.